deepspeed/runtime/utils.py

```diff
@@ -48,6 +48,28 @@ def __init__(self, params):
         self.param_groups.append({'params': params})
 
 
+graph_cache = {}
+
+
+def graph_process(replay_first_step, func, *args, **kwargs):
```

Comment:
XPU also have plan to add graph support.  From current DeepSpeed usage in GPU graph the following three interfaces will be needed:
1. Interface to create accelerator graph object: `_graph_instance = CUDAGraph()` ==> `_graph_instance = get_accelerator().create_graph()`
2. Interface to return a context that will start/end capture data into this graph: `torch.cuda.graph()` ==> `get_accelerator().capture_to_graph(_graph_instance)`
3. Interface to replay a graph: `_graph_instance.replay()` ==> `get_accelerator().replay_graph(_graph_instance)`

With these three interfaces it should be sufficient for use case in DeepSpeed.  Device that does not support graph can return None for create_graph().

@guoyejun FYI on XPU graph support.

Reply:
it's a good interface design, it aligns with the pytorch API, see one example at https://pytorch.org/docs/master/notes/cuda.html#whole-network-capture.

For the XPU/SYCL graph, it is in progress, we'll also align with the current pytorch API for cuda.

Reply:
@delock and @guoyejun, thanks for your comments. I agree generally with your proposal and I am asking our expert also to take a look. I think this PR should use the proposed API to avoid introducing new cuda references into the code as it currently does. It is fine if the API is added to this PR or a separate PR.

Reply:
@inkcherry will add the proposed API along with this PR.  So the API can be verified along with use case.

Reply:
@tjruwase @delock , added the proposed API.

Reply:
I am aligned with these APIs. One potential issue I can see emerging is if there are operations that are permitted under graph capture for only a subset of the supported accelerators. Would an expectation that what is currently viable to graph under CUDA (i.e. no host <-> accelerator copies) be safe to apply as the coding practice for what is placed in graphable regions in DeepSpeed more broadly, or will we need to develop some coding practices around this?

Reply:
the background is that the GPU activities can be captured, but none of the CPU activities. So basically, for the case that the CPU is involved, it is not supported.

btw, the host <-> accelerator copies are supported, see at https://github.com/NVIDIA/cuda-samples/blob/master/Samples/3_CUDA_Features/graphMemoryNodes/graphMemoryNodes.cu#L366C3-L366C26, there are cudaMemcpyAsync for d2h and h2d in the body of function doNegateSquaresInStream.    What not supported is the host <-> accelerator synchronization which involves CPU.

Reply:
@cmikeh2, thanks for your comment. 

I have two thoughts/questions. 

1. Would it work for the new `capture_to_graph()` API to assume the responsibility of validating that the input graph contains only capturable operations for the specific accelerator?  For graphs that are not capturable, `capture_to_graph()` could return `noop_context()` to disable the optimization without breaking functionality. This way, accelerators can independently define/refine the set of capturable operations for graphs. 

2. Is there a way to use the [`make_graphed_callables()`](https://pytorch.org/docs/master/generated/torch.cuda.make_graphed_callables.html#torch.cuda.make_graphed_callables) API. Although, this is currently CUDA specific in torch, perhaps other accelerators could implement similar semantics? This could be used by the `capture_to_graph()` implementations. 

Please share your thoughts if either of this makes sense. Thanks!

Reply:
> 1. Would it work for the new `capture_to_graph()` API to assume the responsibility of validating that the input graph contains only capturable operations for the specific accelerator?  

Do you mean to maintain a list of supported or unsupported torch ops inside `capture_to_graph()`?  I'm afraid this does not work since `capture_to_graph()` just returns a context manager, see more below.

> For graphs that are not capturable, `capture_to_graph()` could return `noop_context()` to disable the optimization without breaking functionality. This way, accelerators can independently define/refine the set of capturable operations for graphs.

When `capture_to_graph()` is called, it goes into the context manager's `__init__` and `__enter__` functions, there's nothing captured at this time, it doesn't know if the following ops will be capturable or not.  See pseudocode below.

```
g = get_accelerator().create_graph()
with get_accelerator().capture_to_graph(g):     # don't know what will be captured at this line
      y = model(input)   # the gpu workload to be captured

...
//fill input with new data
get_accelerator().replay_graph(g)
//fill input with new data
get_accelerator().replay_graph(g)
//fill input with new data
get_accelerator().replay_graph(g)
```

Furthermore, if there's any op in `y = model(input)` above is not capturable, there will be a crash such as `RuntimeError: CUDA error: operation not permitted when stream is capturing`. Looks not feasible for current software stack implementation to use unified code in deepspeed for the graph feature.

Reply:
> 2. Is there a way to use the [`make_graphed_callables()`](https://pytorch.org/docs/master/generated/torch.cuda.make_graphed_callables.html#torch.cuda.make_graphed_callables) API. Although, this is currently CUDA specific in torch, perhaps other accelerators could implement similar semantics? 

I think so that other accelerators will also support `make_graphed_callables()` which is actually based on `torch.cuda.CUDAGraph()` and `torch.cuda.graph`, see https://github.com/pytorch/pytorch/blob/release/2.1/torch/cuda/graphs.py#L354 for more detail.  BTW, I now don't have an answer for where to better use `make_graphed_callables()`.

> This could be used by the `capture_to_graph()` implementations.

My ides is that we need a new accelerator API for `make_graphed_callables`, while keeping the `capture_to_graph()` separated in deepspeed code tree.  The mapping looks like:
torch.cuda.CUDAGraph   ----    get_accelerator().create_graph
with torch.cuda.graph    ----  with get_accelerator().capture_to_graph
torch.cuda.make_graphed_callables  ---- get_accelerator().make_graphed_callables

The connection between `make_graphed_callables` and `graph` happens in the pytorch code, we can keep them separated in deepspeed code, we don't need to maintain a duplicate connection in deepspeed.

Reply:
@guoyejun, thanks so much for your helpful answers. So, it seems that runtime error is the only way to detect invalid capture ops and data? My concern is that a graph capture inside DeepSpeed could succeed on one accelerator but fail (with RuntimeError) on another because of differences in graph support. Do you agree with my concern, @guoyejun?

Reply:
@tjruwase agree with your concern, it makes the code diverged, I now don't have a good answer yet.  btw, it may not be an issue for most cases, I think most of the same graph features will be supported from different vendors.

Reply:
@guoyejun, I guess this is fine for now. Thanks.